irods/client_configuration/__init__.py

```diff
@@ -0,0 +1,154 @@
+    return lambda:getattr(globals()[category], setting)
+
+# #############################################################################
+#
+# Classes for building client configuration categories
+# (irods.client_configuration.data_objects is one such category):
+
+class DataObjects(iRODSConfiguration):
+    __slots__ = ('auto_close',)
+
+    def __init__(self):
+
+        # Setting it in the constructor lets the attribute be a
+        # configurable one and allows a default value of False.
+        #
+        # Running following code will opt in to the the auto-closing
+        # behavior for any data objects opened subsequently.
+        #
+        # >>> import irods.client_configuration as config
+        # >>> irods.client_configuration.data_objects.auto_close = True
+
+        self.auto_close = False
+
+# #############################################################################
+#
+# Instantiations of client-configuration categories:
+
+# The usage "irods.client_configuration.data_objects" reflects the commonly used
+# manager name (session.data_objects) and is thus understood to influence the
+# behavior of data objects.
+#
+# By design, valid configurable targets (e.g. auto_close) are limited to the names
+# listed in the __slots__ member of the category class.
+
+data_objects = DataObjects()
+
+def _var_items(root):
+    if isinstance(root,types.ModuleType):
+        return [(i,v) for i,v in vars(root).items()
+                if isinstance(v,iRODSConfiguration)]
+    if isinstance(root,iRODSConfiguration):
+        return [(i, getattr(root,i)) for i in root.__slots__]
+    return []
+
+DEFAULT_CONFIG_PATH = os.path.expanduser('~/.python_irodsclient')
+
+def save(root = None, string='', file = ''):
+    _file = None
+    auto_close_settings = False
+    try:
+        if not file:
+            file = DEFAULT_CONFIG_PATH
+        if isinstance(file,str):
+            _file = open(file,'w')
+            auto_close_settings = True
+        else:
+            _file = file # assume file-like object if not a string
+        if root is None:
+            root = sys.modules[__name__]
+        for k,v in _var_items(root):
+            dotted_string = string + ("." if string else "") + k
+            if isinstance(v,iRODSConfiguration):
+                save(root = v, string = dotted_string, file = _file)
+            else:
+                print(dotted_string, repr(v), sep='\t\t', file = _file)
+    finally:
+        if _file and auto_close_settings:
+            _file.close()
+
+def load_line(root, setting, value):
+    arr = [_.strip() for _ in setting.split('.')]
+    attr = ''
+    for i in filter(None,arr):
+        if attr:
+            root = getattr(root,attr)
+        attr = i
+    if attr:
+        return setattr(root, attr, ast.literal_eval(value))
+    raise RuntimeError ('bad setting')
+
+_multitab = re.compile('\t+')
+
+_key_value_pattern = re.compile(r'\s*(?P<key>\w+(\.\w+)+)\s+(?P<value>\S.*?)\s*$')
```

Comment:
This feels like a small parser.

Would it be better to use JSON as the representation of the file? I ask because once this ships, it's hard to take it away if users run into issues.

I'm guessing one reason behind this is that users write the options using python-like syntax?

Reply:
For my purposes, this felt better than JSON:
   - as you point out, Pythonic values within JSON might be weird.
   - JSON might imply this is "of a piece" with iRODS, and this is just a way of storing client settings **specific to PRC**
Having said this, we could have multiple loaders, or replace what I've done with JSON, either would be fine. @trel?

Reply:
As long as it doesn't make maintaining it difficult and it's isolated, I think it's fine.

Of course, documentation is necessary for the next person who has to touch it. Consider future Dan six month's to a year having to revisit this for some reason.

Reply:
feels okay.   needs a strong comment and possibly a couple of examples of what it would 'match'.

Reply:
Matching now explained!  Guess we've determined we're not using JSON for this currently, so  ... closing conversation.